### ISSUE TITLE
Transcode loop is indexed by ManifestID

### ISSUE THREAD

**Issue by j0sh** [livepeer/go-livepeer#695]
**Describe the bug**

The ManifestID is broadcaster generated (and may be user supplied). The ManifestID is currently used for two reasons:

* To index streams on the broadcaster for playback
* To index into the orchestrator's transcoding loop

The ManifestID should not be used by the orchestrator in any capacity. There are a few reasons for this:

* Unfavorable behavior within the S3 storage layer with the ManifestID at the root of it all. See [this writeup](https://hackmd.io/0C2Zi0VVSyGEgN_4lEu_WA) for mind-numbing detail on the subject. In short: for O-owned storage, we store source segments in a different location from transcoded results, which generally hinders UX / debugging / forensics / other data collation.
* Minimizes the space of "funny values" supplied by the broadcaster causing issues on the orchestrator (eg, injection). There doesn't seem to be any risk of this now, but doesn't mean there never will be if ManifestID usage proliferates on the orchestrator.
* Orchestrators may be working with multiple broadcasters, increasing the risk of a transcode-loop collision where multiple streams share the same ManifestID. Collisions on the transcoding loop are problematic for a few reasons:
  * The transcode loop itself functions as a backpressure mechanism. If an orchestrator is processing a segment while another comes in, it's likely that the O is unable to keep up with live video. Returning a busy signal here makes the broadcaster fail over elsewhere, depriving the O of legitimate work.
  * Extending the transcode loop beyond its natural expiration also lengthens the time before PM tickets are redeemed.
  * Violating the invariant of "one stream to one transcode loop" may complicate things such as pixel accounting unless we are extra careful.

**Desired Solution**

Collisions and funny values could be mitigated by making any relevant bookkeeping per-broadcaster and hashing the ManifestID. For example, making the transcode loop index `H(<broadcasterAddress> | <manifestID>)` . Even so, this does not address the lingering issues around S3 storage.

Restore the "authentication token" to preserve stateless continuity between GetOrchestrator and segment transcoding. This mitigates all the issues we have on the orchestrator stemming from the Manifest ID.

Here is the commit removing the AuthToken feature: https://github.com/livepeer/go-livepeer/commit/fae70f384ab151b2f2c50495f3c406c706ef0412

The auth token works like this:

```
type AuthToken struct {
  LoopIndex int
  Sig bytes
}
token = AuthToken{ LoopIndex: loopIndex, Sig: Sign(H(loopIndex)) }
```

The token is generated during the GetOrchestrator call. The broadcaster copies it into a header for each segment, which is verified on the orchestrator prior to segment processing.

The drawback is that the AuthToken mechanism is another field and another set of messages to incorporate into the networking protocol. However, the AuthToken is a versatile construct that can be used to incorporate other types of authenticated data should the need arise.

**Comment by yondonfu** [comment]
I like the suggested solution and I have a few questions for clarification:
- Why was the AuthToken originally removed?
- How is `LoopIndex` generated?

**Comment by j0sh** [comment]
> Why was the AuthToken originally removed?

From my recollection of Discord conversations with @ericxtang and @angyangie , we felt that the auth token didn't bring enough value relative to the "surface area" required to implement it.

Although I'm having second thoughts about removing the auth token, I'd also feel better about this if we had more cases with which to use the auth token. Some thoughts:

* Make this workable for offchain broadcasting, where there is currently no authentication. Rather than securing the token with an Ethereum signature, it could be a HMAC.

* Incorporate winProb / faceValue for PM to check against the winProb / faceValue values submitted in the ticket. This might help if the winProb / faceValue is "expensive", eg calculated on a per-broadcaster basis.

> How is `LoopIndex` generated?

This can be any O-selected random value as long as there's no opportunity for B to guess or manipulate the value.

**Comment by yondonfu** [comment]
Leaving a note here just to remind us that once livepeer/go-livepeer#1124 is merged, the manifestID will also be used to index into a list of sessions managed by a load balanced transcoder.

**Comment by j0sh** [comment]
At this point, I'm in favor of resolving this issue via the auth token approach. In addition to the issues outlined above, this would help mitigate some other things that have come up:

* HTTP push clients that retry using different Bs that share the same set of orchestrators currently mean all Bs use the same stream name for a given retry. Coupled with the increased tail latencies we're seeing with https://github.com/livepeer/go-livepeer/pull/1364 , this doesn't give each of B's streams an opportunity to jump to the front of the stack if there's a naming collision; this diminishes the effectiveness of B fallbacks.

* Transcodes on private networks are currently unauthenticated. While the auth token wouldn't resolve the actual authentication of an OrchestratorInfo, it gives a mechanism for the propagation of credentials once we have them. (This was the original impetus for the auth token during the v1 protocol: it avoided a blockchain lookup in the transcoding hot path, and kept orchestrator state minimal.)

When we implement this, I'd also like to resolve all the outstanding issues originally mentioned in this ticket, including unifying the orchestrator's external storage location for source and transcoded segments :

https://github.com/livepeer/go-livepeer/blob/1466f8b10f56427daa2fc66fa23f69d540bc46b4/server/rpc.go#L272

https://github.com/livepeer/go-livepeer/blob/584d38c8a880ea650e59314849cf9a10f08c1c02/core/orchestrator.go#L595

Both can be replaced by the generated transcode loop index, or some derivation of it.

**Comment by yondonfu** [comment]
> HTTP push clients that retry using different Bs that share the same set of orchestrators currently mean all Bs use the same stream name for a given retry. Coupled with the increased tail latencies we're seeing with #1364 , this doesn't give each of B's streams an opportunity to jump to the front of the stack if there's a naming collision; this diminishes the effectiveness of B fallbacks.

Here is a recent example from production that highlights this problem:

```
Sep 13 15:32:19 prod-livepeer-broadcaster-87957886-gqwhs broadcaster: I0913 22:32:18.701299       1 segment_rpc.go:382] Submitting segment nonce=7668967276488037095 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 bytes=41172 orch=https://10.140.103.114:8935 
Sep 13 15:32:19 prod-livepeer-broadcaster-87957886-gqwhs broadcaster: I0913 22:32:18.706089       1 segment_rpc.go:416] Uploaded segment nonce=7668967276488037095 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 orch=https://10.140.103.114:8935 dur=4.73715ms 
Sep 13 15:32:27 prod-livepeer-broadcaster-87957886-gqwhs broadcaster: E0913 22:32:26.701824       1 segment_rpc.go:425] Unable to read response body for segment nonce=7668967276488037095 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 orch=https://10.140.103.114:8935 err=context deadline exceeded 
Sep 13 15:35:07 prod-livepeer-broadcaster-87957886-fhxxj broadcaster: I0913 22:32:31.428797       1 segment_rpc.go:382] Submitting segment nonce=4298253535038596284 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 bytes=41172 orch=https://10.140.103.114:8935 
Sep 13 15:35:07 prod-livepeer-broadcaster-87957886-fhxxj broadcaster: I0913 22:32:31.447854       1 segment_rpc.go:416] Uploaded segment nonce=4298253535038596284 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 orch=https://10.140.103.114:8935 dur=17.701241ms 
Sep 13 15:36:54 prod-livepeer-broadcaster-87957886-6t57z broadcaster: I0913 22:33:11.212098       1 segment_rpc.go:382] Submitting segment nonce=16973090083147076109 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=1 bytes=157920 orch=https://10.140.103.114:8935 
Sep 13 15:36:54 prod-livepeer-broadcaster-87957886-6t57z broadcaster: I0913 22:33:11.232662       1 segment_rpc.go:416] Uploaded segment nonce=16973090083147076109 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=1 orch=https://10.140.103.114:8935 dur=20.51245ms 
Sep 13 15:36:54 prod-livepeer-broadcaster-87957886-6t57z broadcaster: E0913 22:33:11.232715       1 segment_rpc.go:448] Transcode failed for segment nonce=16973090083147076109 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=1 orch=https://10.140.103.114:8935 err=OrchestratorBusy 
Sep 13 15:37:00 prod-livepeer-broadcaster-87957886-fhxxj broadcaster: E0913 22:32:39.441600       1 segment_rpc.go:425] Unable to read response body for segment nonce=4298253535038596284 manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 seqNo=2 orch=https://10.140.103.114:8935 err=context deadline exceeded 
```

`prod-livepeer-broadcaster-87957886-6t57z` encounters an `OrchestratorBusy` error. I confirmed that the segment submission to that particular O was the first one for manifestID=5eda7871-e5ac-42cc-9150-e03291149d03 from that B. So, B encounters this error before it even submits a segment to that O! We can see that prior to that another B had sent a segment to that same O for the same manifestID. That B encountered a timeout error which implies that the O was probably still working on the segment. So, the HTTP push client probably retried with `prod-livepeer-broadcaster-87957886-6t57z`, but `prod-livepeer-broadcaster-87957886-6t57z` also sent the segment to the same O and since it was still working on a segment for that manifestID it returned an `OrchestratorBusy` error. It would've been preferable for the O to at the very least use a different session when `prod-livepeer-broadcaster-87957886-6t57z` sent its segment (although it'd probably be most ideal for that B to use a completely different O altogether).

**Comment by yondonfu** [comment]
Another issue that would be resolved by the removing the use of B supplied manifestIDs for indexing into data structures on Os:

At the moment, Bs & Os keep track of stream balances using a map of the form ETH address -> manifestID -> balance. This map structure can cause issues when multiple Bs use the same ETH address and/or when multiple Os use the same ETH address. 

In the former case, each B will have its own view of its stream balance with O. But, the O will update the stream balance for *all* Bs with the same ETH address. In the latter case, each O will have its own view of its stream balance with B. But, the B will update the stream balance for *all* Os with the same ETH address. In both cases, B might send payments thinking its stream balance is X when O sees the stream balance as Y s.t. Y < X due to payment exchange with other Bs using the same ETH address. This can result in an "Insufficient balance" error being returned by O.

If O generates an identifier during discovery that is used by both B & O and that identifier is used in the stream balance map I think the above scenario could be avoided.
